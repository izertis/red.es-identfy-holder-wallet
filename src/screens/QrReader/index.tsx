import { useState, useEffect, useContext } from 'react'
import { StatusBar } from 'react-native'
import QrCustomMarker from './components/QrCustomMarker'
import { useTranslation } from 'react-i18next'
import { SCREEN } from '../../constants/screens'
import qrReaderI18nKeys from './i18n/keys'
import { NavigationInjectedProps } from 'react-navigation'
import QrReaderStyled from './styles'
import { BarCodeScanner } from 'expo-barcode-scanner'
import { ActivityIndicator } from 'react-native-paper'
import { ColorKeys, getThemeColor } from '../../constants/Colors'
import { useMatchingID } from '../../services/lacchain/didMatch'
import { MessageContext } from '../../context/UserMessage.context'
import { useNavigation } from '@react-navigation/native'
import { useOpenId } from '../../services/open-id'

interface Props extends NavigationInjectedProps {
  issuer?: string
}

const QrReader = (props: Props) => {
  const { handleCredentialOffer } = useOpenId()
  const { handleMatchingID } = useMatchingID()
  const { t } = useTranslation(SCREEN.QrReader)
  const [isLoading, setIsLoading] = useState<boolean>(true)
  const [hasPermission, setHasPermission] = useState<boolean | null>(null)
  const [isReading, setIsReading] = useState<boolean>(false)
  const [scanned, setScanned] = useState<boolean>(false)
  const { showMessage } = useContext(MessageContext)
  const navigation = useNavigation()

  useEffect(() => {
    initializeQrReader()
  }, [])

  const initializeQrReader = async () => {
    await checkCameraPermission()
    setIsLoading(false)
  }

  const checkCameraPermission = async () => {
    const { status } = await BarCodeScanner.getPermissionsAsync()
    setHasPermission(status === 'granted')
    if (status === 'undetermined') await requestCameraPermission()
  }

  const requestCameraPermission = async () => {
    const { status } = await BarCodeScanner.requestPermissionsAsync()
    setHasPermission(status === 'granted')
  }

  const handleBarCodeScanned = async ({ data }: any) => {
    // Verify here if QR code corresponds to ebsi, lacchain or alastria
    if (!scanned) {
      setScanned(true)
      setIsReading(true)
      setIsLoading(true)
      try {
        if (data.startsWith("openid")) {
          await handleCredentialOffer(data)

        } else if (data.startsWith("matching-id")) {
          await handleMatchingID(data)
          showMessage({ content: t?.(qrReaderI18nKeys.SUCCESS_ASOCIATE_IDENTIFIER), type: "success" })

        } else {
          try {
            // Data could also be a json object (openId flow)
            const parsedData = JSON.parse(data)
            await handleCredentialOffer(JSON.stringify(parsedData))

          } catch (error) {
            showMessage({ content: t?.(qrReaderI18nKeys.ERROR), type: "error" })
          }
        }
      } catch (error: any) {
        showMessage({ content: error.message, type: "error" })

      } finally {
        setIsReading(false)
        setIsLoading(false)
        setTimeout(() => {
          if (navigation.canGoBack()) navigation.goBack()
        }, 3000)
      }
    }
  }

  if (isLoading) {
    return (
      <QrReaderStyled.Container>
        <ActivityIndicator
          style={{ position: "absolute", alignSelf: "center", top: "45%", zIndex: 100000 }}
          color={getThemeColor(ColorKeys.primary)}
          size={70}
        />
      </QrReaderStyled.Container>
    )
  } else if (hasPermission) {
    return (
      <QrReaderStyled.Container>
        <StatusBar barStyle="light-content" />
        <QrReaderStyled.QrScanner
          onBarCodeScanned={!isReading ? handleBarCodeScanned : () => { }}
        />
        <QrCustomMarker
          isReading={isReading}
          subtitle={t(qrReaderI18nKeys.SCAN_QR, {
            issuer: props.issuer || t(qrReaderI18nKeys.ISSUER),
          })}
          navigation={props.navigation}
        />
      </QrReaderStyled.Container>
    )
  } else {
    return (
      <QrReaderStyled.Container style={{ justifyContent: 'center' }}>
        <StatusBar backgroundColor={'black'} barStyle="light-content" />
        <QrReaderStyled.RequestingText>
          {t(qrReaderI18nKeys.PERMISSIONS_MESSAGE)}
        </QrReaderStyled.RequestingText>
      </QrReaderStyled.Container>
    )
  }
}

export default QrReader
