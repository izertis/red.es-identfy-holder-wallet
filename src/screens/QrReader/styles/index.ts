import defaultStyles from "./defaultStyles";
import proyectSyles from "./proyectStyles";

const QrReaderStyled = {
  ...defaultStyles,
  ...proyectSyles
};

export default QrReaderStyled;
