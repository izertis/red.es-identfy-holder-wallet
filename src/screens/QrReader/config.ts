import i18next from "../../../i18n.config"
import { SCREEN } from "../../constants/screens"
import { navigationScreensConfigType } from "../../types/navigation"
import QrReader from "./"
import es from "./i18n/es"

const ScreenName = SCREEN.QrReader

i18next.addResourceBundle("es", ScreenName, es)

const QrReaderConfig: navigationScreensConfigType = {
  name: ScreenName,
  component: QrReader,
  options: {
    headerShown: false,
  },
}

export default QrReaderConfig
