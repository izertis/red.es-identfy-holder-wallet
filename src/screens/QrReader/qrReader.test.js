import React from "react";
import { render } from "@testing-library/react-native";

import QrReader from ".";
import { useNavigation } from "@react-navigation/native";

describe("<QrReader />", () => {
  it("renders correctly", () => {
    const navigation = { navigate: jest.fn() };
    const result = render(<QrReader navigation={navigation} />);
		expect(result.toJSON()).toMatchSnapshot()
  });
});
