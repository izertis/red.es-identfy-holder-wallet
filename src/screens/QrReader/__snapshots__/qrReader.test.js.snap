// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QrReader /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "black",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="progressbar"
    accessibilityState={
      Object {
        "busy": true,
      }
    }
    accessible={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "alignSelf": "center",
          "position": "absolute",
          "top": "45%",
          "zIndex": 100000,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "height": 70,
          "opacity": 1,
          "width": 70,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 70,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 70,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 35,
                "overflow": "hidden",
                "width": 70,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 70,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 70,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 35,
                    "overflow": "hidden",
                    "width": 70,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "#eb5b5b",
                      "borderRadius": 35,
                      "borderWidth": 7,
                      "height": 70,
                      "width": 70,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 70,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 70,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 35,
                "overflow": "hidden",
                "top": 35,
                "width": 70,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 70,
                  "transform": Array [
                    Object {
                      "translateY": -35,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 70,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 35,
                    "overflow": "hidden",
                    "width": 70,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "#eb5b5b",
                      "borderRadius": 35,
                      "borderWidth": 7,
                      "height": 70,
                      "width": 70,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
